### ISSUE TITLE
[Perf][B300] Tile-parallel cached decoding for GSA and Raven

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/flash-linear-attention#12]
## Problem

`fused_recurrent_gsa` uses a monolithic inference kernel for cached single-token decode. At the shipped GSA and Raven defaults (`hidden_size=2048`, `num_heads=4`, `num_slots=64`, `expand_k=expand_v=1`), each head has `K=V=512` and `M=64`.

For `B=1`, the current grid launches only four CTAs. Each CTA serially walks eight 64-wide K-state tiles and eight 64-wide V-state tiles while updating the two FP32 recurrent states. This is a credible B300 wave-underfill bottleneck on every cached token across the default 24-layer GSA and Raven models, but it is an unmeasured hypothesis until the profile-first gate below passes.

## Proposed optimization

Add a bounded two-stage inference path for forward cached decode only:

1. A K-state stage with grid `(ceil(K / BK), B * H)` updates disjoint K-state rows and writes FP32 partial slot logits `[B, H, NK, M]`.
2. A V-state stage with grid `(ceil(V / BV), B * H)` reads those partials in ascending K-tile order, performs the existing FP32 slot reduction and softmax, updates disjoint V-state columns, and writes the corresponding output columns.

At the public default this changes four long-running CTAs into 32 CTAs per stage while adding only about 8 KiB of partial logits at `B=1, H=4, NK=8, M=64`.

The optimized dispatch is deliberately restricted to:

- `T == 1`, no autograd, forward decode;
- `NG == 1`;
- `output_final_state=True` with valid initial recurrent states;
- `cu_seqlens is None` and `reverse=False`;
- profile-proven large `K/V/M` buckets.

Keep the existing monolithic kernel byte-for-byte reachable as the fallback for GQA, training, varlen, reverse mode, missing final-state output, small state sizes, unsupported shapes, and measured losing buckets. Do not change public APIs, model/config files, state dtype, fast-math flags, training forward/backward, or numerical tolerances.

Each optimized program must exclusively own its state/output slice. Preserve the current ascending K-tile accumulation order in stage 2, mask non-multiple `K/V/M` tails, and use `tl.int64` for program/grid-derived bases, strides, and offsets. The only new workspace is the compact FP32 partial-logit tensor; do not add atomics or state-sized scratch.

## Scope

Expected files:

- `fla/ops/gsa/fused_recurrent.py`
- `tests/ops/test_gsa.py` in an oracle/fixture-only commit before production edits, then frozen
- `benchmarks/ops/registry.py` for a forward-only `fused_recurrent_gsa_decode` entry
- `benchmarks/ops/run.py` only to make the documented `skip_backward=True` path avoid constructing or calling backward during phase-1 warmup

The runner correction must include a focused regression proving that a forward-only entry never invokes backward in warmup or measurement. No general runner refactor is in scope.

## Fixture-first correctness gate

Start the feature branch from exact upstream baseline `de09abdfd6ba09df58907bb4f2637b902297dd2a`.

Before changing production code:

1. Add a deterministic default-shape inference fixture: `B=1`, `T=63`, `HQ=H=4`, `K=V=512`, `M=64`, BF16 q/k/v/s/g, FP32 recurrent states, and token-by-token cached `fused_recurrent_gsa(..., output_final_state=True)` compared with `naive_recurrent_gsa`.
2. Add the registry entry and the narrow `skip_backward` runner regression.
3. Run the new fixture against untouched production code. If it is red, OOR, or needs a reference/tolerance/dtype/numeric change, record no-go before kernel work.
4. Commit this evidence plumbing as immutable `BENCH_BASE`, then freeze `tests/ops/test_gsa.py`, `fla/ops/gsa/naive.py`, tolerances, inputs, and numeric flags.

The existing full GSA oracle remains mandatory and covers recurrent inference, GQA fallback, dense/varlen training, forward/backward, and state gradients under allocator NaN poisoning.

Promotion gates include:

```bash
python -m pytest -q tests/ops/test_gsa.py
python -m benchmarks.ops.verify --op fused_recurrent_gsa_decode --base <BENCH_BASE> --modes fwd
python -m pytest -q tests/models/test_modeling_gsa.py
python -m pytest -q tests/models/test_modeling_raven.py
```

Also run dependent tests discovered for `fla/ops/gsa/fused_recurrent.py`, compile/lint hooks required by the repository, and both Raven generation variants. No test, reference, tolerance, or warning-suppression edit is allowed after the fixture is frozen.

## Profile-first B300 gate

Before production edits, profile the exact default decode endpoint on an isolated B300 allocation with backend dispatch controlled and recorded. Measure the inference kernel's share of steady cached-token layer/model CUDA time, the four-CTA wave underfill, and allocation/launch overhead.

Stop without source optimization if:

- the inference kernel is less than 10% of the relevant steady decode endpoint;
- the baseline does not show the expected underfill/resource signature;
- a proof split kernel cannot compile without material spills/resource collapse; or
- the summed two-stage time cannot plausibly repay the added launch and compact-buffer allocation.

Underfill is the target mechanism; ordinary host launch overhead is not itself a reason to reject the candidate unless the added launch erases the gain.

## Repository-native benchmark plan

Register forward-only, non-grad `T=1` shapes with FP32 states and final-state output:

- public defaults: `B={1,2,4,8}`, `H=4`, `K=V=512`, `M=64`;
- adjacent guards: `H={4,8}`, `K/V={128,256,512}`, `M={32,64}`;
- at least one asymmetric `K != V` case;
- explicit fallback controls for GQA and small states.

Compare production HEAD against immutable `BENCH_BASE` in the same B300 allocation with warmed caches and at least five interleaved rounds. Report per-shape median, mean, standard deviation, min, p10, p20, p80, p90, sample count, and equal-weight geomean. Forward-only native timing must include the compact temporary allocation as exercised by the public wrapper.

For user-path evidence, instantiate deterministic BF16 default GSA and Raven models without network checkpoints, prefill once, then time 256-512 steady cached single-token steps at `B=1/2/4`. Use the actual model cache path and report tokens/s distributions, clocks, idle evidence, environment, exact commands, and Slurm job IDs.

## NCU evidence

Collect baseline and candidate reports for the exact default shape using:

- `--set full` plus PM sampling/warp-state sections;
- `--set source` plus `SourceCounters`.

Compare summed kernel duration, grid/waves, active and eligible warps, achieved occupancy, registers, spills/local traffic, DRAM/L2 bytes and throughput, SM issue/throughput, and dominant source stalls. The expected explanation is four severely underfilled baseline CTAs versus 32 CTAs per candidate stage without a spill or traffic penalty that erases the benefit. Store raw reports outside git and keep exact environment/hash/job metadata.

## Acceptance thresholds

- Full frozen op and GSA/Raven model gates pass unchanged.
- Default `B=1,H=4,K=V=512,M=64` kernel latency improves at least 1.30x median.
- `B=1/2/4` public-default kernel geomean improves at least 1.20x.
- Default GSA cached-token throughput improves at least 8% at `B=1`.
- Default Raven cached-token throughput improves at least 5% at `B=1`.
- No registered adjacent, `B=2/4`, or fallback endpoint regresses more than 3% outside measured noise.
- NCU corroborates higher tile parallelism without material spill/occupancy collapse.

If the extra launch, compact allocation, or repeated slot softmax erases the endpoint gain, record no-go rather than widening into training, model code, another backend, or a shape-specific lookup table.

## Duplicate and mergeability audit

At upstream `de09abdf`, no open issue, PR, or remote branch targets GSA/Raven cached-decode tile parallelism or `fused_recurrent_gsa_inference_kernel`. Nearby open work concerns unrelated generation correctness and short convolution. Historical GSA changes are correctness/API/addressing work, not this ownership/grid rewrite.

This is a localized two-stage Triton inference optimization shared by two shipped model families, with an unchanged fallback and no API/model-format change. It is independently mergeable if the frozen gates and public-path thresholds pass.

**Comment by heiheiha798** [comment]
## B300 evaluation result: no-go

The bounded two-stage cached-decode prototype was correctness-safe but did not meet the public endpoint performance thresholds on B300, so this investigation stops without a commit or draft PR.

### Reproducible state

- Source baseline: `de09abdfd6ba09df58907bb4f2637b902297dd2a`
- Frozen fixture/BENCH_BASE and branch HEAD: `ae6282c7b5c3584252412038570051ed2f38c14a`
- Branch: `b300/gsa-tile-parallel-cached-decode`
- Device: NVIDIA B300 SXM6 AC, SM103
- Candidate validation: Slurm job `5276`, completed with exit `0:0` in `00:28:30`

### Correctness

The complete frozen command ran before benchmarking:

```text
python -m benchmarks.ops.verify --op fused_recurrent_gsa_decode --base ae6282c7b5c3584252412038570051ed2f38c14a --modes fwd
```

`tests/ops/test_gsa.py` passed `31/31` tests with 14 warnings under the repository's NaN-poisoned test environment. This includes the frozen BF16 token-by-token cache-parity fixture and all existing dense, varlen, state, output, and gradient coverage. There was no correctness failure.

### Same-session performance

The eligible BF16 `K=V=512, M=64, NG=1` shapes compared the dirty candidate directly with BENCH_BASE in the same allocation:

| Shape | Median speedup |
| --- | ---: |
| B1 / H4 | 0.98x |
| B2 / H4 | 0.98x |
| B4 / H4 | 0.93x |
| B8 / H4 | 0.93x |
| B1 / H8 control | 0.95x |

The B1/B2/B4/B8 equal-weight geomean is about `0.95x`, versus the required `>=1.20x`; B1 is `0.98x`, versus the required `>=1.30x`. B4 and B8 also exceed the allowed 3% regression. Guarded fallback controls were approximately neutral at `1.00x`.

Baseline profiling in jobs `5266` and `5269` had confirmed that the monolithic kernel was material and severely underfilled. The proof nevertheless shows that two launches, compact partial allocation, and repeated finalization cost more than the extra grid parallelism saves at the public endpoints. The explicit stop rule therefore applies before model integration or candidate NCU; neither was run, and no speedup claim is supported.

### Artifacts and disposition

- Frozen fixture gate: `profile/fused_recurrent_gsa_decode-opt/fixture-5257.log`
- Baseline timing/profile: `profile/fused_recurrent_gsa_decode-opt/profile-baseline-5266.log`
- Replacement Raven timing and baseline NCU: `profile/fused_recurrent_gsa_decode-opt/profile-replacement-5269.log`
- Candidate gate and same-session comparison: `profile/fused_recurrent_gsa_decode-opt/candidate-verify-5276.log`
- Iteration record: `profile/fused_recurrent_gsa_decode-opt/OPT_LOG.md`

The verifier's JSON was job-local temporary output and is not retained; the durable job log contains the full environment, gate, warmup, comparison tables, clocks, and final idle state. The two-stage source prototype remains uncommitted in the preserved cycle-11 migration worktree at `fla/ops/gsa/fused_recurrent.py` (142 inserted lines) for forensic reproduction only.

Conclusion: **performance no-go**. This is not an environment or correctness failure, but the core mechanism is below threshold and regresses target shapes, so there is no bounded salvage candidate. The issue remains open as the investigation record.
